packages/react-query-next-experimental/src/HydrationStreamProvider.tsx

```diff
@@ -0,0 +1,192 @@
+   *    - This means that we might have some new entries in the cache that needs to be flushed
+   *    - We pass these to the client by inserting a `<script>`-tag where we do `window[id].push(serializedVersionOfCache)`
+   * 2. (Happens in browser) In `useEffect()`:
+   *   - We check if `window[id]` is set to an array and call `push()` on all the entries which will call `onEntries()` with the new entries
+   *   - We replace `window[id]` with a `push()`-method that will be called whenever new entries are received
+   **/
+  function UseClientHydrationStreamProvider(props: {
+    children: React.ReactNode
+    /**
+     * Optional transformer to serialize/deserialize the data
+     * Example devalue, superjson et al
+     */
+    transformer?: DataTransformer
+    /**
+     * **Client method**
+     * Called in the browser when new entries are received
+     */
+    onEntries: (entries: TShape[]) => void
+    /**
+     * **Server method**
+     * onFlush is called on the server when the cache is flushed
+     */
+    onFlush?: () => TShape[]
+  }) {
+    // unique id for the cache provider
+    const id = `_${React.useId()}`
+    const idJSON = JSON.stringify(id)
+
+    const [transformer] = React.useState(
+      () =>
+        (props.transformer ?? {
+          // noop
+          serialize: (obj: any) => obj,
+          deserialize: (obj: any) => obj,
+        }) as unknown as TypedDataTransformer<TShape>,
+    )
+
+    // <server stuff>
+    const [stream] = React.useState<TShape[]>(() => {
+      if (typeof window !== 'undefined') {
+        return {
+          push() {
+            // no-op on the client
+          },
+        } as unknown as TShape[]
+      }
+      return []
+    })
+    const count = React.useRef(0)
+    const onDehydrateRef = React.useRef(props.onFlush)
+    onDehydrateRef.current = props.onFlush
+    useServerInsertedHTML(() => {
+      // This only happens on the server
+      stream.push(...(onDehydrateRef.current?.() ?? []))
+
+      if (!stream.length) {
+        return null
+      }
+      // console.log(`pushing ${stream.length} entries`)
+      const serializedCacheArgs = stream
+        .map((entry) => transformer.serialize(entry))
+        .map((entry) => JSON.stringify(entry))
+        .join(',')
+
+      // Flush stream
+      stream.length = 0
+
+      const html: string[] = [
+        `window[${idJSON}] = window[${idJSON}] || [];`,
+        `window[${idJSON}].push(${serializedCacheArgs});`,
+      ]
+      return (
+        <script
+          key={count.current++}
+          dangerouslySetInnerHTML={{
+            __html: html.join(''),
+          }}
+        />
+      )
+    })
+    // </server stuff>
+
+    // <client stuff>
+    const onEntriesRef = React.useRef(props.onEntries)
+    onEntriesRef.current = props.onEntries
+
+    // Client: consume cache:
+    const onEntries = React.useCallback(
```

Comment:
since `onEntries` is only used once (in the effect), we could move it inside the effect to avoid `useCallback`, which means the dependencies of the effect would become `[id, transformer]`

Reply:
Sounds good to me!

Reply:
I prefer inlining as much as possible, I've had that debate countless times lol

Reply:
inlined here: 73367547